### PR TITLE
Limit  to the bounds of MapView. (#1226)

### PR DESCRIPTION
**Pull request by pengdev**
### Summary of changes

Port https://github.com/mapbox/mapbox-maps-android/pull/1226 to v10.5 release branch

## Pull request checklist:
 - [x] Briefly describe the changes in this PR.
 - [ ] Include before/after visuals or gifs if this PR includes visual changes.
    
 - [x] Write tests for all new functionality. If tests were not written, please explain why.
 - [ ] Optimize code for java consumption (`@JvmOverloads`, `@file:JvmName`, etc).
 - [ ] Add example if relevant.
 - [ ] Document any changes to public APIs.
 - [ ] Run `make update-api` to update generated api files, if there's public API changes, otherwise the `verify-api-*` CI steps might fail.
 - [ ] Update [CHANGELOG.md](../CHANGELOG.md) or use the label 'skip changelog', otherwise `check changelog` CI step will fail.
 - [x] If this PR is a `v10.[version]` release branch fix / enhancement, merge it to `main` firstly and then port to `v10.[version]` release branch.

Fixes: < Link to related issues that will be fixed by this pull request, if they exist >

PRs must be submitted under the terms of our Contributor License Agreement [CLA](https://github.com/mapbox/mapbox-maps-android/blob/main/CONTRIBUTING.md#contributor-license-agreement).

### CODE REVIEW

File: sdk/src/test/java/com/mapbox/maps/MapboxMapTest.kt [hunk 1]
```diff
@@ -1027,4 +1026,79 @@ class MapboxMapTest {
       styleObserver.setLoadStyleListener(options, any(), any())
     }
   }
+}
+
+@RunWith(ParameterizedRobolectricTestRunner::class)
```
Comment (alexander-kulikovskii) [inline sdk/src/test/java/com/mapbox/maps/MapboxMapTest.kt:1031]: is it necessary to use `Robolectric` here?

File: sdk/src/main/java/com/mapbox/maps/MapboxMap.kt [hunk 2]
```diff
@@ -740,13 +740,22 @@ class MapboxMap :
...
    *
-   * @return Returns a screen coordinate on the screen in [MapOptions.size] platform pixels.
+   * @return Returns a screen coordinate on the screen in [MapOptions.size] platform pixels. If the screen coordinate is outside of the bounds of [MapView] the returned screen coordinate contains -1 for both coordinates.
    */
   override fun pixelForCoordinate(coordinate: Point): ScreenCoordinate {
     checkNativeMap("pixelForCoordinate")
-    return nativeMap.pixelForCoordinate(coordinate)
+    val coordinate = nativeMap.pixelForCoordinate(coordinate)
+    val screenSize = nativeMap.size
+    return if (coordinate.x in 0.0..screenSize.width.toDouble() && coordinate.y in 0.0..screenSize.height.toDouble()) {
```
Comment (alexander-kulikovskii) [inline sdk/src/main/java/com/mapbox/maps/MapboxMap.kt:754]: btw, do you know, creating range here will be converted to `0.0 <= coordinate.x <= screenSize.width.toDouble()` or not?
maybe it can be not super effective?

File: sdk/src/test/java/com/mapbox/maps/MapboxMapTest.kt (hunk 1 above)
Comment (pengdev) [reply sdk/src/test/java/com/mapbox/maps/MapboxMapTest.kt:1031]: yes, to use shadow class to bypass the Logger that goes through gl-native

File: sdk/src/main/java/com/mapbox/maps/MapboxMap.kt (hunk 2 above)
Comment (pengdev) [reply sdk/src/main/java/com/mapbox/maps/MapboxMap.kt:754]: we have parameterized tests to cover the corner cases here.

Comment (alexander-kulikovskii) [reply sdk/src/main/java/com/mapbox/maps/MapboxMap.kt:754]: tests is good, but here I'm talking about performance

File: sdk/src/test/java/com/mapbox/maps/MapboxMapTest.kt (hunk 1 above)
Comment (alexander-kulikovskii) [reply sdk/src/test/java/com/mapbox/maps/MapboxMapTest.kt:1031]: okay

File: sdk/src/main/java/com/mapbox/maps/MapboxMap.kt (hunk 2 above)
Comment (pengdev) [reply sdk/src/main/java/com/mapbox/maps/MapboxMap.kt:754]: It's also suggested in https://kotlinlang.org/api/latest/jvm/stdlib/kotlin.ranges/range-to.html.

According to the [source](https://github.com/JetBrains/kotlin/blob/6a670dc5f38fc73eb01d754d8f7c158ae0176ceb/libraries/stdlib/src/kotlin/ranges/Ranges.kt#L64-L97)

```
override fun contains(value: Double): Boolean = value >= _start && value <= _endInclusive
```

So I think it should be efficient to use.

**Review by alexander-kulikovskii — APPROVED** [review]

**Review by kiryldz — APPROVED** [review]
